bcnmy/bundler
# Feature: Frontrunning Protection & necessary refactors

# 📖 Context
## Type of change

Why? We don't know yet how Flashbots RPC is going to behave in production & the production config needs to be updated with the Flashbots URL.

## Why are we doing this?

For frontrunning:
- We were losing money on gas intensively because of frontrunning on Eth Mainnet for the previous couple of days
- If the relayer balance gets drained the relayer goes out of service
- If all relayers go out of service, the network goes out of service

For refactoring:
- The code was spaghetti so I had to do a lot of refactoring on the spot, though I don't like doing features and refactors in the same PR, I had no choice.
- The services were booting up too slow so I had to do something about it because debugging was taking forever.

## What did we do?

For frontrunning:
- Use the Flashbots RPC for relaying transactions to the EntryPoint contract
- Use the regular provider for other call, especially gas estimation because the Flashbots RPCs don't return any revert reasons, so it's impossible to estimate properly or debug what's going on
- Implement [nonce management for Flashbots](https://docs.flashbots.net/flashbots-protect/nonce-management)

For refactoring:
- All necessary refactoring to understand what's going on and properly implement everything written above.
- Refactor the **Service Manager** so it loads chains in parallel instead of sequentially. 
- Deleted or commented some broken legacy tests

## How Has This Been Tested?

- All the internal transactions you can see sent from [my smart account](https://etherscan.io/address/0xb7744Ce158A09183F01296bCB262bB9F65E2FFd7#internaltx) in the last 3 days.

# 👀 How do I review this?

- Check for any obvious mistakes in the code, I know it's a lot but this is important.
- The `EVMNetworkService` had the most changes, take a look at it.
- Check changes in `default.json`.
- 🚨 Check the Service Manager changes and see if there's any obvious things that can go wrong.
- Send some Ethereum Mainnet user ops to the production bundler after it's deployed
